### PR TITLE
Skip linux-bridge filtering if no "bridge" section

### PR DESCRIPTION
**Pull request by qinqon**
**Is this a BUG FIX or a FEATURE ?**:
We have found that at some envs the "bridge" section from linux-bridge
is not present in case the bridge is down, this change ignore the
filtering in this is the case to preveng a segmentation fault.

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind enhancement

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/nmstatectl/remote.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
+package nmstatectl
```
Comment (phoracek) [inline pkg/nmstatectl/remote.go:1]: This should not be a part of this PR, right?

Comment (qinqon) [reply pkg/nmstatectl/remote.go]: Removed, this is from the PoC.

File: pkg/state/filter.go [hunk 2]
```diff
@@ -58,9 +58,24 @@ func filterOutDynamicAttributes(iface map[string]interface{}) {
 		return
 	}
 
-	bridge := iface["bridge"].(map[string]interface{})
+	bridgeIface, ok := iface["bridge"]
```
Comment (phoracek) [inline pkg/state/filter.go:61]: use `hasBridge` instead of `ok`. Then the following condition makes more sense.

Comment (phoracek) [inline pkg/state/filter.go:61]: How is this `iface`? We just know that it has the type of an empty interface. Maybe `bridgeRaw` would be slightly better?

File: pkg/state/filter_test.go [hunk 3]
```diff
@@ -307,4 +307,66 @@ routes:
...
+  state: down
+  ipv4:
+    enabled: false
+  ipv6:
+    enabled: false
+  lldp:
+    enabled: false
+  mac-address: A2:EE:84:7B:42:4C
+  mtu: 1500
+routes:
```
Comment (phoracek) [inline pkg/state/filter_test.go:328]: Please drop everything that is not necessary for this test. We do not care about routes, do we?

**Review by phoracek — CHANGES_REQUESTED** [review]

File: pkg/state/filter_test.go [hunk 4]
```diff
@@ -307,4 +307,66 @@ routes:
 			Expect(returnedState).To(MatchYAML(filteredState))
 		})
 	})
+	Context("when there is a linux bridge without 'bridge' options because is down", func() {
+		BeforeEach(func() {
+			state = nmstate.NewState(`
+interfaces:
+- name: eth1
```
Comment (rhrazdil) [inline pkg/state/filter_test.go:314]: the `ethernet` interface seems not relevant for the test case either

File: pkg/state/filter_test.go [hunk 5]
```diff
@@ -307,4 +307,44 @@ routes:
...
+  type: linux-bridge
+  state: down
+  ipv4:
+    enabled: false
+  ipv6:
+    enabled: false
+  lldp:
+    enabled: false
+  mac-address: A2:EE:84:7B:42:4C
+  mtu: 1500
```
Comment (phoracek) [inline pkg/state/filter_test.go:324]: None of this is needed to test that you fixed the issue, is it?

Comment (qinqon) [reply pkg/state/filter_test.go:324]: I have replicate what we have in the env that is failing the bridge is like this.

File: pkg/state/filter.go [hunk 6]
```diff
@@ -30,7 +30,11 @@ func FilterOut(currentState shared.State) (shared.State, error) {
 }
 
 func filterOutRoutes(kind string, state map[string]interface{}, interfacesFilterGlob glob.Glob) {
-	routes := state["routes"].(map[string]interface{})
+	routesRaw, hasRoutes := state["routes"]
+	if !hasRoutes {
+		return
+	}
+	routes := routesRaw.(map[string]interface{})
```
Comment (phoracek) [inline pkg/state/filter.go:37]: Why don't you test whether the casting went right here, but you do check it further down?

**Review by phoracek — CHANGES_REQUESTED** [review]

File: pkg/state/filter.go (hunk 6 above)
Comment (qinqon) [reply pkg/state/filter.go:37]: In the case of routes there is always "config" and "running"

File: pkg/state/filter_test.go (hunk 5 above)
Comment (phoracek) [reply pkg/state/filter_test.go:324]: But then it is not a test for missing bridge. It is a test for missing bridge *and* another issues that we haven't thought about and may be in the code. We may catch a bug by chance here, but it feels really random and non-fitting for this specific test. You may argue that we should include snapshot of the whole state of a very complicated node to catch more potential issues

Comment (qinqon) [reply pkg/state/filter_test.go:324]: But we want to test with sane data right ? this attributes are present at bridge like this when a bridge is down.

File: pkg/state/filter.go (hunk 6 above)
Comment (phoracek) [reply pkg/state/filter.go:37]: How is it relevant to casting? Do you mean that in case of "bridge" for instance we can get nil?

Comment (qinqon) [reply pkg/state/filter.go]: Ahh right, misunderstood, added it.

File: pkg/state/filter_test.go (hunk 5 above)
Comment (qinqon) [reply pkg/state/filter_test.go]: Ok, will just keep name, type and state.

Comment (phoracek) [reply pkg/state/filter_test.go]: You are writting a unit test. The test should be concerned only about a specific subpart of the system. If we need the whole state to be populated to get one specific function running, then we have an issue in our architecture and end-up with ton of unmaintainable yamls.

We had a bug with missing bridge. You should write a minimal reproducer and then a fix.

If you expect that e.g. "lldp" may affect run of the function, then we need a test for invalid lldp.

**Review by phoracek — COMMENTED** [review]
/approve

Looks good. Would you lgtm @rhrazdil if all seems fine to you?

**Review by rhrazdil — COMMENTED** [review]
/lgtm

**Comment by enp0s3** [comment]
@phoracek @qinqon sorry but it's not clear to me what do you mean by "bridge is down" - was that administratively down or was the link of the bridge down? Perhaps I am not familiar with the states

**Pull request merged**